### DISCUSSION TITLE
Modular Application with Identity gives 500 error

### DISCUSSION THREAD

**Discussion by upaelfid** [OrchardCMS/OrchardCore#15351] [Q&A] [answered]
Im creating a modular application. When I reference the module without using Microsoft.AspNetCore.Identity.UI I can see the resources that the referenced module uses. But when I add Microsoft.AspNetCore.Identity.UI then I get 500 errors for all the resources that are referenced by the module. When I go inspect the element that is not being served I get:

An unhandled exception occurred while processing the request.
ArgumentException: Value cannot be null or empty. (Parameter 'areaName')
Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapAreaControllerRoute(IEndpointRouteBuilder endpoints, string name, string areaName, string pattern, object defaults, object constraints, object dataTokens)

Any help would be greatly appreciated

**Comment by ns8482e** [comment]
The razor class library with area is not supported with orchard core modular framework #4247 #10230 #10287

**Comment by upaelfid** [comment]
Thank you Niraj! Do you know if there a way to add identity to my module without the cms?

On Feb 17, 2024, at 8:13 AM, Niraj Soni ***@***.***> wrote:

﻿

The razor class library with area is not supported with orchard core modular framework

**Comment by ns8482e** [comment]
I have not tried but you could implement `IApplicationFeatureProvider<ViewsFeature>` for Identity Assembly  and register in your Program.cs

**Reply by upaelfid** [reply]
Not having any luck yet. I am creating a module that has an [Authorize] attribute on one of its controllers, this way a user has to sign in to edit an item. Do you know of a way of getting around the razor class library issue?

**Reply by ns8482e** [reply]
I think the issue might be Identity area not matching with assembly.

Try this---
- add Microsoft identity ui reference to mvc web
- create a mvc module named "Identity" and add reference to mvc web
- Run app

**Reply by upaelfid** [reply]
Thank you so much for responding! Here are the steps I am taking.
-I create a .net core web app mvc
-A controller called CoreImg
-A view called index. The view has .jpg image
-I add OrchardCore.Module.Targets as a reference

-I create a project called Website
-Add OrchardCore.Application.Mvc.targets
-set up the Program.cs with AddOrchardCore().AddMvc();
and the useOrchardCore();

when I go to https://localhost/CoreImage/Index I see the .jp image

If I rewrite everything as above but for the Website part, I add the Identity framework, and add [Authorize] to the CoreImg controller... When I got to https://localhost/CoreImage/Index  it asks me to login with my credentials. I log in but now the Image isnt being served and when I inspect it it gives me a 500 error

**Reply by ns8482e** [reply]
500 error might be due to missing authorization in your orchard pipeline

**Comment by ns8482e** [comment]
@upaelfid  This is how I made it to work.

1. Create a dummy module `Identity`
```cli
 dotnet new ocmodulemvc -o Identity
```

2. Create a Modular module `Module1`. Create your controller here in this module with `[Authorize]` attribute.

```cli
dotnet new ocmodulemvc -o Module1
```
3. Create a ASP.NET Core web application ( from VS or using following command line)  and add project references and package references as below.
```cli
dotnet new ocmvc -o ModularWebApp
```
```xml
  <ItemGroup>
    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="8.0.2" />
    <PackageReference Include="Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore" Version="8.0.2" />
    <PackageReference Include="Microsoft.AspNetCore.Identity.EntityFrameworkCore" Version="8.0.2" />
    <PackageReference Include="Microsoft.AspNetCore.Identity.UI" Version="8.0.2" />
    <PackageReference Include="OrchardCore.Application.Mvc.Targets" Version="1.8.2" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="8.0.2" />
  </ItemGroup>

  <ItemGroup>
    <ProjectReference Include="..\Identity\Identity.csproj" />
    <ProjectReference Include="..\Module1\Module1.csproj" />
  </ItemGroup>
```
5.  Update `Program.cs` as following

```cs
using Microsoft.AspNetCore.Identity;
using Microsoft.EntityFrameworkCore;

var builder = WebApplication.CreateBuilder(args);

builder.Services
    .AddOrchardCore()
    .AddMvc()
     // Orchard Specific Pipeline
     .ConfigureServices(services =>
     {
		 // Add services to the container.
		 var connectionString = builder.Configuration.GetConnectionString("DefaultConnection") ?? throw new InvalidOperationException("Connection string 'DefaultConnection' not found.");
		 services.AddDbContext<ApplicationDbContext>(options =>
			 options.UseSqlServer(connectionString));
		 services.AddDatabaseDeveloperPageExceptionFilter();

		 services.AddDefaultIdentity<IdentityUser>(options => options.SignIn.RequireConfirmedAccount = true)
			 .AddEntityFrameworkStores<ApplicationDbContext>();

	 })
     .Configure((app, routes, services) =>
     {
		 app.UseAuthorization();
	})
;

var app = builder.Build();

if (!app.Environment.IsDevelopment())
{
    app.UseExceptionHandler("/Error");
    // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
    app.UseHsts();
}

app.UseHttpsRedirection();

app.UseStaticFiles();

app.UseOrchardCore();

app.Run();
```

**Reply by upaelfid** [reply]
Thank you so much for giving me this direction. Ive been wrestiling trying to duplicate this. Do you add the scaffolded Identity item to the dummy "Identiy" module or do you add it to the ModularWebApp?

**Reply by ns8482e** [reply, marked as answer]
I used OC `Identity` module which as nothing in it. It allows OC to be aware of `Identity` module with area `Identity`. 
`Identity` area is already exists in  Identity UI, so OC will serve UI from Identity UI.

All Identity related Data are in main WebApp

![image](https://github.com/OrchardCMS/OrchardCore/assets/23270244/fced276e-71dc-4517-b646-ecd7bb0ab03c)

![image](https://github.com/OrchardCMS/OrchardCore/assets/23270244/6f1baa1d-ab9b-4f06-a294-b9b6be82065c)

**Reply by upaelfid** [reply]
That works like MAGIC!
